lib/ansible/modules/cloud/vmware/vmware_guest.py

```diff
@@ -1919,7 +1919,11 @@ def select_datastore(self, vm_obj=None):
             # TODO: really use the datastore for newly created disks
             if 'autoselect_datastore' in self.params['disk'][0] and self.params['disk'][0]['autoselect_datastore']:
                 datastores = self.cache.get_all_objs(self.content, [vim.Datastore])
-                datastores = [x for x in datastores if self.cache.get_parent_datacenter(x).name == self.params['datacenter']]
+                if self.params['esxi_hostname']:
+                    # Filtering datastores to those one attached to esxi host selected (it could be local storage)
+                    datastores = [x for x in datastores if self.params['esxi_hostname'].split(".",1)[0] in [h.key.config.network.netStackInstance[0].dnsConfig.hostName for h in x.host]]
```

Comment:
You ignore the FQDN, and this can lead to some unwanted situation. E.g: you can have two VM with the same hosts with a different FQDN. e.g: esxi1.prod and esxi1.testing
Is it possible to compare the FQDN instead?

Reply:
You are right, 

Adding FQDN on the Datastore filtering.